### ISSUE TITLE
Bug with digital signatures?

### ISSUE THREAD

**Issue by michaelhyolee** [gettalong/hexapdf#27]
I have an existing digitally-signed PDF:
![image](https://user-images.githubusercontent.com/13178209/37312500-a82d7a1a-2621-11e8-83c9-12dd2fcc3b48.png)

I then open the file using HexaPDF, then `write` with `optimize: false`. A simple `open` then `write` with no modifications done to the PDF.

I expected a PDF with valid signatures, like the original, but instead got this:

![image](https://user-images.githubusercontent.com/13178209/37313721-e37cc31e-2626-11e8-81f8-9f2f09d2da98.png)

Any help would be appreciated!

**Comment by gettalong** [comment]
Tl;dr: Digital signatures are not yet supported with HexaPDF.

If you load and then write a PDF, the resulting blob is not the same as the original blob due to differences in serialization, ordering, etc. Therefore the signature must be renewed when writing, but since HexaPDF doesn't yet support digital signatures, no new signature is written and the old signature objects stay in the PDF, resulting in your "invalid signature".

Another way to get around this would be the incremental writing feature of PDF which would put the original PDF as a blob in the resulting PDF (no changes = no signature invalidation) and just append the changes. This would technically already work with HexaPDF but is also not yet implemented as there was no need yet. This technique is often used by PDF viewers to save a PDF form with the new data without "changing" the original PDF.

So you have currently two options: Use another library/a post-processing step if digital signatures are required. Or remove the signature objects from the PDF before writing if only the "invalid signature" indication in viewers should go away.

**Comment by michaelhyolee** [comment]
Thank you for replying so quickly!

The incremental writing solution is really what I was looking for. We have an existing PDF with a valid signature. We want to add text/images to that PDF in a new revision without invalidating the existing signature for the previous revision.

Would this be a large or complex change to HexaPDF?

**Comment by gettalong** [comment]
I have had a solution for the incremental writing in my mind for quite a while now but there was no need for it, yet.

It basically boils down to two things:

* Making sure that all changed parts get placed into a new Revision object.
* Changing the Writer class to be able to only write out one revision on top of an existing document.

I'm currently working on finishing the basics for document layout but after that I can have a look how complex the needed changes would really be.

**Comment by michaelhyolee** [comment]
Amazing, please keep us in the loop!

**Comment by gettalong** [comment]
@michaelhyolee Just a heads-up: Since I have released the latest changes two days ago, I'm currently working on incremental writing support. I will have a working (but probably not fully optimized) version by next weekend.

**Comment by gettalong** [comment]
@michaelhyolee I have pushed the latest changes to the repository. Those changes include the incremental writing feature. I have tested it with a signed PDF and when using the incremental writing mode the signature stayed valid (tested with the `pdfsig` binary).

I will close this this issue now since it should be resolved. If you still have questions or if the solution doesn't work for you, I will gladly reopen the issue.
